### ISSUE TITLE
torch.var and torch.std are not compatible with np.var and np.std

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#50010]
torch.var and torch.std have the following signatures:

```
torch.var(input, dim, unbiased=True, keepdim=False, *, out=None)
torch.std(input, dim, unbiased=True, keepdim=False, *, out=None)
```

Which are incompatible with NumPy's signatures:

```
numpy.var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=<no value>)
numpy.std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=<no value>)
```

Because PyTorch's default of `unbiased=True` is equivalent to a NumPy default of `ddof=1`. 

PyTorch's use of the term "unbiased" is also problematic. While Bessel's correction (ddof=1) does produce an unbiased estimator of the variance, it DOES NOT produce an unbiased estimator of the standard deviation. See [this wiki article](https://en.wikipedia.org/wiki/Unbiased_estimation_of_standard_deviation) for a discussion.

I propose we deprecate PyTorch's default behavior by doing the following:

- [ ] adding a ddof kwarg
- [ ] warning when that kwarg is not set for a release (this requires auditing and updating PyTorch's internal callsites to avoid warnings when people use PyTorch)
- [ ] requiring the kwarg be set (this probably requires writing an upgrader for serialized torchscript programs)
- [ ] removing the "unbiased" arg and adding a default ddof value of zero

The net effect of this will be that PyTorch's var and std will lose "unbiased," gain "ddof," and change their default behavior to be the same as NumPy's.

**Comment by mruberry** [comment]
Triage review for socialization.

**Comment by rgommers** [comment]
A couple of comments:

- Pandas uses `ddof=1` as default, NumPy uses `ddof=0`. PyTorch default corresponds to `ddof=1`:

```
>>> import torch                                                                                                                         
>>> t = torch.rand(20)                                                                                                                   
>>> torch.std(t)                                                                                                                         
tensor(0.3483)
>>> np.std(t.numpy())                                                                                                                    
0.33949822
>>> np.std(t.numpy(), ddof=1)                                                                                                            
0.34831783
```

- Other array/tensor libraries all use `ddof=0`, except TensorFlow which doesn't have any keyword like this: https://github.com/data-apis/array-api-comparison/blob/main/signatures/statistics/std.md
- We may be stuck with it in NumPy et al., but I personally dislike `ddof` as a name because I've seen enough users misunderstand it as "degrees of freedom" rather than "delta degrees of freedom". 
- This is one of the few cases where in the array API standardization work we chose to deviate from NumPy, using `correction` as a more descriptive name (IIRC copied from some other language), see https://github.com/data-apis/array-api/pull/17#discussion_r471839878. Not finalized, but a data point to consider.

What they mean (from the `numpy.std` docstring): `ddof=1` provides an unbiased estimator of the variance of the infinite population. `ddof=0` provides a maximum likelihood estimate of the variance for normally distributed variables

**Comment by mruberry** [comment]
I agree with you, @rgommers, that the name is not great. It is certainly not common. See, for example, Wikipedia's discussion of Bessel's correction [here](https://en.wikipedia.org/wiki/Bessel%27s_correction). `correction` sounds like a fine alternatively. Luckily if we match the default value then changing the kwargs name shouldn't be a problem. 

It's true, by the way, that those `ddof` values do that for variance, but I don't believe they do for the standard deviation (see linked Wiki article above). The actual effect of the value is just to modify the term `n/(n - ddof)`.

**Comment by mattip-q** [comment]
In #51953 I started to try to deal with `torch.rand(2, 2, 1).var(axis=2)` producing `nan`s where numpy produces `0`s. Since this PR touches the same code, they will probably conflict.

**Comment by heitorschueroff** [comment]
With the new Python Array API Standard, we should replace unbiased with `correction=0`, not `ddof`. I created an issue here https://github.com/pytorch/pytorch/issues/61492 to capture this.

**Comment by peterbell10** [comment]
The API already uses `correction` and not `ddof`, although the documentation won't be updated before #56398 is merged.

**Comment by mruberry** [comment]
FYI that PR did hit some issues with BC/FC, and we need to review it more carefully. @heitorschueroff has setup a discussion to get us started with the torchscript and PyTorch mobile teams.

**Comment by soulitzer** [comment]
From triage review:
- It is probably not worth it given the long process and large churn. 
- The default is already documented in the docs, but we can make it clearer that it differs from numpy
- There is an Array API compatibility layer that can help here.
